### PR TITLE
MARKET-1674 [Navigation Dots Component]

### PR DESCRIPTION
**Pull request by Fialkaksu**
The Navigation Dots сomponent is a navigation element designed for use in web applications. It provides a visually appealing way to navigate through different sections on a webpage.

[MARKET-1674](https://bugs.backendless.com/browse/MARKET-1674)
[UI-Builder](https://develop.backendless.com/app/TestComponentsApp/ui-builder/NavigationDots/pages/landing)
[Desktop Preview](https://populardowntown.backendless.app/api/files/ui-builder/containers/NavigationDots/index.html?page=landing)

image

### CODE REVIEW

File: components/bl-navigation-dots/README.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+| Tooltip <br> `tooltip`                      | *Checkbox*                                     | `true`            |                          | NO           | YES        | Controls whether a tooltip will be shown when hovering over a dot by default or not.                                      |
+| Tooltip Background <br> `tooltipBackground` | *Color*                                        | `rgba(0,0,0,0.7)` | Tooltip Background Logic | YES          | YES        | This logic allows you to specify the background of the tooltip.                                                           |
+| Tooltip Color <br> `tooltipColor`           | *Color*                                        | `#ffffff`         | Tooltip Color Logic      | YES          | YES        | This logic allows you to specify the color of the tooltip.                                                                |
+
+## Events
+
+| Name     | Triggers                               | Context Blocks           |
+|----------|----------------------------------------|--------------------------|
+| On Click | when the user clicks on the component. | activeSection: `Element` |
```
Comment (serhmel) [inline components/bl-navigation-dots/README.md:25]: `activeSection `-> `Active Section` ?

File: components/bl-navigation-dots/src/index.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+      if (sectionElement) {
+        const rect = sectionElement.getBoundingClientRect();
+
+        if (rect.top <= windowHeight / 2 && rect.bottom >= windowHeight / 2) {
+          newActiveDotIndex = i;
+          break;
+        }
+      }
+    }
```
Comment (v-excelsior) [inline components/bl-navigation-dots/src/index.js:42]: there should be a function that will make it more readable

File: components/bl-navigation-dots/src/index.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    const sectionElement = document.getElementById(anchor);
+
+    if (sectionElement) {
+      sectionElement.scrollIntoView({ behavior: 'smooth' });
+      onClick({ activeSection: sections[index] });
+    }
+
+    setTimeout(() => {
+      setScrollEnabled(true);
+    }, 1000);
```
Comment (v-excelsior) [inline components/bl-navigation-dots/src/index.js:65]: could be one lined

File: components/bl-navigation-dots/src/index.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        { sections?.map((section, index) => (
+          <li key={ index }>
+            <a
+              className={ cn('nav-item', dotShape, { active: activeDotIndex === index }) }
+              href={ `#${ section.anchor }` }
+              onClick={ () => dotClickHandler(index) }>
+
+              { tooltip && (
+                <span className="nav-item-tooltip">{ section.anchor }</span>
+              ) }
```
Comment (v-excelsior) [inline components/bl-navigation-dots/src/index.js:102]: one line?

File: components/bl-navigation-dots/src/index.js [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+            </a>
+          </li>
```
Comment (v-excelsior) [inline components/bl-navigation-dots/src/index.js:105]: Component?

File: components/bl-navigation-dots/styles/index.less [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    position: relative;
+    align-items: center;
+    outline: none;
+    text-decoration: none;
+    padding: @bl-customComponent-navigation-dots-nav-item-padding;
+
+    &:before {
+      content: "";
+      display: block;
+      pointer-events: auto !important;
```
Comment (v-excelsior) [inline components/bl-navigation-dots/styles/index.less:66]: Why here is an important?

File: components/bl-navigation-dots/src/index.js [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+  if (!display) {
+    return null;
+  }
+
+  return (
+    <div className={ cn('bl-customComponent-navigation-dots', classList) } ref={ elRef } style={ styles }>
+      <ul className="nav">
+
+        { sections?.map((section, index) => (
```
Comment (v-excelsior) [inline components/bl-navigation-dots/src/index.js:93]: How about add `if` statement that will check for sections and print warning if they are invalid?

File: components/bl-navigation-dots/README.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+<p align="center">
+  <img src="./thumbnail.png" alt="main thumbnail" width="780"/>
+</p>
+
+## Properties
+
+| Property                                    | Type                                           | Default value     | Logic                    | Data Binding | UI Setting | Description                                                                                                               |
+|---------------------------------------------|------------------------------------------------|-------------------|--------------------------|--------------|------------|---------------------------------------------------------------------------------------------------------------------------|
+| Sections <br> `sections`                    | *JSON*                                         | `[]`              | Sections Logic           | YES          | NO         | This logic allows you to specify the array of sections.  Signature of sections: list of DOM elements `[Element, Element]` |
```
Comment (v-excelsior) [inline components/bl-navigation-dots/README.md:13]: I'm not clearly understand how we could create dynamic count of sections using data binding?

File: components/bl-navigation-dots/README.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+| Tooltip <br> `tooltip`                      | *Checkbox*                                     | `true`            |                          | NO           | YES        | Controls whether a tooltip will be shown when hovering over a dot by default or not.                                      |
+| Tooltip Background <br> `tooltipBackground` | *Color*                                        | `rgba(0,0,0,0.7)` | Tooltip Background Logic | YES          | YES        | This logic allows you to specify the background of the tooltip.                                                           |
+| Tooltip Color <br> `tooltipColor`           | *Color*                                        | `#ffffff`         | Tooltip Color Logic      | YES          | YES        | This logic allows you to specify the color of the tooltip.                                                                |
+
+## Events
+
+| Name     | Triggers                               | Context Blocks            |
+|----------|----------------------------------------|---------------------------|
+| On Click | when the user clicks on the component. | Active Section: `Element` |
```
Comment (v-excelsior) [inline components/bl-navigation-dots/README.md:25]: On Section Change event?

File: components/bl-navigation-dots/src/index.js [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+        newActiveDotIndex = i;
+        break;
+      }
+    }
+  }
+
+  return newActiveDotIndex;
+}
+
+function anchorValidation(anchors) {
```
Comment (v-excelsior) [inline components/bl-navigation-dots/src/index.js:163]: anchorValidation -> validateAnchors

File: components/bl-navigation-dots/src/index.js [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+  if (anchors && !anchors.every(anchor => anchor && typeof anchor === 'string')) {
+      console.error("Error: anchor in 'anchors' prop is missing, invalid, or contains non-string value in NavigationDotsComponent.");
```
Comment (v-excelsior) [inline components/bl-navigation-dots/src/index.js:165]: Not quite clear which anchor is missing. How about validate them one by one, and if some is invalid - print it with index in array?

File: components/bl-navigation-dots/src/index.js [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+
+    const activeAnchor = anchors[index];
+    const sectionElement = document.getElementById(activeAnchor);
+
+    if (sectionElement) {
+      sectionElement.scrollIntoView({ behavior: scrollBehavior });
+      onAnchorChange({ activeAnchor });
+    }
+
+    setTimeout(() => setScrollEnabled(true), 1000);
```
Comment (v-excelsior) [inline components/bl-navigation-dots/src/index.js:60]: Clean timeouts?

File: components/bl-navigation-dots/src/index.js [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+
+  useEffect(() => {
+    window.addEventListener('scroll', scrollHandler);
+
+    return () => window.removeEventListener('scroll', scrollHandler);
+  }, [scrollHandler]);
+
+  useEffect(() => {
+    anchorValidation(anchors);
+  }, [anchors]);
```
Comment (v-excelsior) [inline components/bl-navigation-dots/src/index.js:75]: one line

File: components/bl-navigation-dots/src/index.js [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+    }
+  };
+
+  if (!display) {
+    return null;
+  }
+
+  return (
+    <div className={ cn('bl-customComponent-navigation-dots', classList) } ref={ elRef } style={ styles }>
+
```
Comment (v-excelsior) [inline components/bl-navigation-dots/src/index.js:92]: redundant line

File: components/bl-navigation-dots/src/index.js [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+      { anchors &&
+        <Nav
+          anchors={ anchors }
+          activeDotIndex={ activeDotIndex }
+          dotShape={ dotShape }
+          tooltip={ tooltip }
+          dotClickHandler={ dotClickHandler }
+        />
+      }
```
Comment (v-excelsior) [inline components/bl-navigation-dots/src/index.js:101]:
```suggestion
      { anchors && (
        <Nav
          anchors={ anchors }
          activeDotIndex={ activeDotIndex }
          dotShape={ dotShape }
          tooltip={ tooltip }
          dotClickHandler={ dotClickHandler }
        />
      )}
```

File: components/bl-navigation-dots/src/index.js [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+  return (
+    <div className={ cn('bl-customComponent-navigation-dots', classList) } ref={ elRef } style={ styles }>
+
+      { anchors &&
+        <Nav
+          anchors={ anchors }
+          activeDotIndex={ activeDotIndex }
+          dotShape={ dotShape }
+          tooltip={ tooltip }
+          dotClickHandler={ dotClickHandler }
```
Comment (v-excelsior) [inline components/bl-navigation-dots/src/index.js:99]: dotClickHandler -> onItemClick

File: components/bl-navigation-dots/src/index.js [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+
+    if (sectionElement) {
+      const rect = sectionElement.getBoundingClientRect();
+
+      if (rect.top <= windowHeight / 2 && rect.bottom >= windowHeight / 2) {
+        newActiveDotIndex = i;
+        break;
+      }
+    }
+  }
```
Comment (v-excelsior) [inline components/bl-navigation-dots/src/index.js:158]: Array.find?

**Review by v-excelsior — CHANGES_REQUESTED** [review]

**Pull request merged**